### DISCUSSION TITLE
How to use getters and setters in AdonisV5?

### DISCUSSION THREAD

**Discussion by brunodsazevedo** [adonisjs/core#2745] [Help]
I used Adonis in version 4.1, and I remember that there were getters and setters methods for dealing with requests and responses. But in version 5, I didn't find anything referring to these methods.

My intention would be to create a get method with a function to create a hash for my primary-keys from my table to serve my front end, and when I was going to do a query or update, the hash provided earlier would be decrypted by the setter method.

Any tips on how to proceed with this type of situation? I tried to work with computed() methods but I confess that I had difficulty implementing it even with the example in the documentation.

**Comment by RomainLanz** [comment]
Hey @brunodsazevedo! 👋 

You can use the `consume` and `prepare` options inside the `@column()` decorator.
📚 https://docs.adonisjs.com/reference/orm/decorators#column

Example: https://twitter.com/romainlanz/status/1283365113119019008

**Reply by Vishwas-verma** [reply]
Further way you can use hooks too like this-

**Set**

```
  @beforeSave()
  public static async encryptKey(model: Model) {
       model.PrimaryKey = await Encryption.encrypt(model.PrimaryKey)
  }
```
**Get** 

Like this you can make after save,after find,after fetch,after create hooks based on your condtion and decrypt value in them the other way would be using computed function-
```
@computed
  public get decryptKey() {
    const key: any = await Encryption.decrypt(this.primaryKey); 
    return key;
  }
```

**Reply by brunodsazevedo** [reply]
> Hey @brunodsazevedo! wave
> 
> You can use the `consume` and `prepare` options inside the `@column()` decorator.
> books https://docs.adonisjs.com/reference/orm/decorators#column
> 
> Example: https://twitter.com/romainlanz/status/1283365113119019008

Hello @RomainLanz . I tried the method in question, but without success. This was the answer I got from the method:

```
{
  "message": "E_RUNTIME_EXCEPTION: \"Encryption.decrypt\" expects a string value",
  "stack": "Exception: E_RUNTIME_EXCEPTION: \"Encryption.decrypt\" expects a string value\n    at Encryption.decrypt (/home/bruno/projetos/wacash/backend/node_modules/@adonisjs/encryption/build/src/Encryption/index.js:92:19)\n    at Object.consume (/home/bruno/projetos/wacash/backend/app/Models/User.ts:20:44)\n    at /home/bruno/projetos/wacash/backend/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:1113:37\n    at Array.forEach (<anonymous>)\n    at Proxy.$consumeAdapterResult (/home/bruno/projetos/wacash/backend/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:1100:40)\n    at Function.$createFromAdapterResult (/home/bruno/projetos/wacash/backend/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:171:18)\n    at /home/bruno/projetos/wacash/backend/node_modules/@adonisjs/lucid/build/src/Orm/QueryBuilder/index.js:118:50\n    at Array.reduce (<anonymous>)\n    at ModelQueryBuilder.execQuery (/home/bruno/projetos/wacash/backend/node_modules/@adonisjs/lucid/build/src/Orm/QueryBuilder/index.js:116:37)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
  "code": "E_RUNTIME_EXCEPTION"
}
```

This is my @column:
```
@column({
    isPrimary: true,
    prepare: (value: number) => Encryption.encrypt(value),
    consume: (value: string) => Encryption.decrypt(value),
  })
  public id: number
```

**Reply by brunodsazevedo** [reply]
> Further way you can use hooks too like this-
> 
> **Set**
> 
> ```
>   @beforeSave()
>   public static async encryptKey(model: Model) {
>        model.PrimaryKey = await Encryption.encrypt(model.PrimaryKey)
>   }
> ```
> 
> **Get**
> 
> Like this you can make after save,after find,after fetch,after create hooks based on your condtion and decrypt value in them the other way would be using computed function-
> 
> ```
> @computed
>   public get decryptKey() {
>     const key: any = await Encryption.decrypt(this.primaryKey); 
>     return key;
>   }
> ```

@Vishwas-verma , The method actually worked, but I expected another result. Instead of returning a new hashed object, I expected my "id" key to be replaced with my hash.
